### PR TITLE
[NewDot API] Add No api side effect actions rule

### PR DESCRIPTION
**Pull request by mountiny**
cc @yuwenmemon 

Adding a new rule to make sure that no function inside `/actions/` files has two API/action method calls which would imply side effects.

**Description from the design doc:**

>no-api-side-effect-actions
>This rule will check to see if a method is calling API.write or API.read and then also calling another method that is defined by a file in the /actions/ directory. If so, it is violation. It will be enforceable in all files in the /actions/ directory.

Fixes https://github.com/Expensify/Expensify/issues/211244

### Tests

Added automated jest tests and made sure they pass.

### CODE REVIEW

**Comment by mountiny** [comment]
@marcaaron @roryabraham Asking you two for a review as you are the most familiar with this repository.

This PR is still on HOLD, however, it can be reviewed already whenever you will get around to it 🙌

**Comment by mountiny** [comment]
`no-multiple-api-calls` only checks if there are 2 calls to API or DeprecatedAPI.

In this one we also check if there is an API call and some function defined in `/actions/` called as both cannot be called from one function. I guess these rules could be combined, but for atomicity and clearer guidance when the error is thrown keeping two would be better imho.

File: eslint-plugin-expensify/CONST.js [hunk 1]
```diff
@@ -14,5 +14,6 @@ module.exports = {
 
         // eslint-disable-next-line max-len
         NO_MULTIPLE_API_CALLS: 'Do not call API or DeprecatedAPI multiple times in the same method. The API response should return all the necessary data in a single request, and API calls should not be chained together.',
+        NO_API_SIDE_EFFECTS_ACTIONS: 'Avoid actions with side effects. Each action should have only one API call or one action method call.',
```
Comment (marcaaron) [inline eslint-plugin-expensify/CONST.js:17]: "side effects" feels like code for some more specific thing that we are trying to avoid.

>Each action should have only one API call or one action method call.

So an action can either call one API or another action method? Why?

**Comment by marcaaron** [comment]
Left a couple of comments. I am not really sure which situation we are trying to prevent with this rule. Maybe we can clarify that on the original issue before continuing here. I read the explanation in the design doc and issue, but I don't 100% get it.

**Comment by mountiny** [comment]
@marcaaron @roryabraham Thanks for chiming in. @yuwenmemon is now surely bombarded with many threads on the NewDot API after coming back from ooo, but I think. he can explain this best.

What I have understood from this design doc and other Offline first is that we are aiming to have 1:1 structure with API calls so if there is a defined action in the `/actions/` directory, it should make only one call to `API` _or_ call another action, for which the same applies so it can essentially always be just one API command per action method. Since we do not want to chain API calls in the action methods.

But I must admit I am not completely sure about the entire context so I would like to discuss with Yuwen 🙌

File: eslint-plugin-expensify/CONST.js (hunk 1 above)
Comment (mountiny) [reply eslint-plugin-expensify/CONST.js:17]: I have struggled a lot with formulating the error message and I am very open to changing that to something clearer which I believe we can come down to after discussing this rule.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            filename: './src/libs/notActions/Test.js',
+        },
+    ],
+    invalid: [
+        {
+            code: `
+                import Action from './src/libs/actions/Action.js'
+                function test() {
+                    API.write('Report_AddComment', params);
+                    Action.perform(params);
```
[the remark this answers is not in the document]
Comment (mountiny) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:62]: I _think_ the Action would have another API call inside (or another action but eventually with API call in the end) so technically this would be two API calls in one function/method which is the camouflaged `no-multiple-api-calls` rule.

**Comment by marcaaron** [comment]
>Since we do not want to chain API calls in the action methods.

Maybe we can go through a case which is a concern and why we do not want to do them?
I didn't see any examples of chained API calls in the tests for this so maybe those would also be good to add?

**Comment by mountiny** [comment]
> I didn't see any examples of chained API calls in the tests for this so maybe those would also be good to add?

Good call, I will add one!

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                function test() {
+                    API.write('Report_AddComment', params).then((value) => {
+                        Action.perform(params);
+                    });
+                }
+            `,
+            errors: [{
+                message,
+            }],
+            filename: './src/libs/actions/Test.js',
```
Comment (mountiny) [inline eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:83]: Invalid chaining test.

Comment (marcaaron) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:83]: Doesn't really make sense though as `API.write()` does not return a promise.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:62]: Ok making more sense now, thanks! So the assumption is that `Action.something()` has an `API.write()` or `API.read()` in it because it (1) is a function and (2) it is exported from the `/actions` directory. 

Do we check this or guarantee that the function is also calling `API.write()` or `API.read()` or is that just implied by 1 and 2 ? 

Seems like there are actions that do not call API at all. Should we let them be called or not? How about things that set Onyx data? Should those be allowed to be called?

I think the change is fine just pushing to get the guidance a bit clearer.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        },
+        {
+            code: `
+                import Action from './src/libs/actions/Action.js'
+                function test() {
+                    Action.perform(params);
+                }
+            `,
+            filename: './src/libs/actions/Test.js',
```
Comment (marcaaron) [inline eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:29]: In this example we are importing an action and wrapping it in a function. Maybe this should not be allowed as you should call the action directly not wrap it in another action 🤔

File: eslint-plugin-expensify/CONST.js (hunk 1 above)
Comment (marcaaron) [reply eslint-plugin-expensify/CONST.js:17]: >Avoid actions with side effects.

I guess I am hung up on what makes it a "side effect". We don't return a promise from the other methods (only `makeRequestWithSideEffects()`) so they can't actually have any unless they use that method.

>Each action should have only one API call or one action method call.

This seems like the more important thing, but feels like it should also explain a bit more about why we enforce this.

**Comment by marcaaron** [comment]
Oh also maybe it is for a separate PR, but I think a common case we will see is people trying to use both `Onyx.merge()` and `API.write()` in the same "action" method. We shouldn't allow it because it should be handled in the data options.

**Comment by yuwenmemon** [comment]
> I read the explanation in the design doc and issue, but I don't 100% get it.

@marcaaron Perhaps for more context you can refer to the conversation here: https://expensify.slack.com/archives/C01GTK53T8Q/p1650899039398159?thread_ts=1650386234.773169&cid=C01GTK53T8Q

Essentially, "It would be an anti-pattern for one action to call another action, because that would imply that calling one action isn't enough."

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:62]: @marcaaron for an example you could look at `validateLogin` here: https://github.com/Expensify/App/blob/0b5f84330063763de41896d5fe74eacf432fa9c3/src/libs/actions/User.js#L222

We validate the login and then call the action to get UserDetails, which goes ahead and calls another API command. This should instead be done in one single API command.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 3 above)
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:83]: Yeah, I think you would want to use `read` in this test case.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        },
+        {
+            // Test that two functions in one file with different API/action calls are valid
```
Comment (yuwenmemon) [inline eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:32]: I love the comments like this where we describe what we're testing for in plain English, can we add these to all test cases?

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 3 above)
Comment (mountiny) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:83]: Good points. I haven't really thought about how valid the examples are since the linter just goes through the tokens and searches for keywords.

I agree it will be better with real use example haha 😄

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 4 above)
Comment (mountiny) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:29]: I am not sure if I understand what you mean here. Do you mean we should have `Action()`? I think we currently get the actions in such a way but I agree that will probably be refactored soon, for example of [Inbox.requestInboxCall()](https://github.com/Expensify/App/blob/51fb61808d84f4aae4c6cbb3f6875d43d355fbb5/src/pages/RequestCallPage.js#L153).

If we will have strictly one action file - one method, then I think the rule you suggest would make a lot of sense. Did I understand it right?

**Comment by mountiny** [comment]
Addressed the comments, thanks for the feedback. Ready for another look.

File: eslint-plugin-expensify/CONST.js [hunk 6]
```diff
@@ -14,5 +14,6 @@ module.exports = {
 
         // eslint-disable-next-line max-len
         NO_MULTIPLE_API_CALLS: 'Do not call API or DeprecatedAPI multiple times in the same method. The API response should return all the necessary data in a single request, and API calls should not be chained together.',
+        NO_API_SIDE_EFFECTS_ACTIONS: 'Avoid calling actions within other actions. If an action needs to call another action consider combining the two actions as a singular API call instead.',
```
Comment (marcaaron) [inline eslint-plugin-expensify/CONST.js:17]:
```suggestion
        NO_API_SIDE_EFFECTS_ACTIONS: 'Calling actions from inside other actions is forbidden. If an action needs to call another action combine the two actions into a singular API call instead.',
```

Just a tad more forceful. We don't ever want to allow this.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+const message = require('../CONST').MESSAGE.NO_API_SIDE_EFFECTS_ACTIONS;
+
+const ruleTester = new RuleTester({
+    parserOptions: {
+        ecmaVersion: 6,
+        sourceType: 'module',
+    },
+});
+
+ruleTester.run('no-api-side-effect-actions', rule, {
```
Comment (marcaaron) [inline eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:12]: It would be more direct to call the rule something like `no-call-actions-from-actions` or `prefer-one-one-one-design`.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 4 above)
Comment (marcaaron) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:31]: **clarification**: the example in the test case is not something anyone would ever do - yet implied as "valid"
**suggestion:** we should remove it unless we can think of a real world example where someone would need to create a function that wraps an action.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                    API.write('Report_AddComment', params);
+                    Action.perform(params);
+                }
+            `,
+            filename: './src/libs/notActions/Test.js',
+        },
+    ],
+    invalid: [
+        {
+            // Making an API call and calling an Action shuold be invalid
```
Comment (marcaaron) [inline eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:60]:
```suggestion
            // Making an API call and calling an Action should be invalid
```

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:65]: @yuwenmemon Ok, so a good "invalid" example of chaining would be something like:

```js
function validateAndGetDetails(params) {
    API.makeRequestWithSideEffects('ValidateEmail', params)
        .then((response) => {
            API.write('GetUserDetails', response);
        });
}
```

As we can't chain with `write` or `read` yea?

[the remark this answers is not in the document]
Comment (marcaaron) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:65]: >Do we plan to make sure that each action//API call has its own js file in /actions/ in which there would be only one function and that one function is exported

At the moment an "action method" is very loose but in general is something that either:

- Sets data in Onyx
- Calls an API and also sets data in Onyx

There are unfortunate actions that do things like format data, but those are not really actions at all and just colocated around actions. All of that kind of stuff I would expect will light up with a lot of errors once this change is introduced.

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 7 above)
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:12]: `no-call-actions-from-actions` sounds good to me!

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js:65]: Yeah I see what you mean.

> So this rule would be more helpful if it didn't target old code (e.g. DeprecatedAPI usages are fine but API usages are not fine). Not sure if that is possible - but I'd think we want to avoid people trying to "fix" old deprecated code when those rules only apply to the refactors we are planning or new APIs being added.

I like this. I think it's possible because we can just exclude any action that is making a call to `DeprecatedAPI` right? @mountiny do you think that would be possible?

[the remark this answers is not in the document]
Comment (mountiny) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js]: @marcaaron @yuwenmemon That makes sense to leave `DeprecatedAPI` out.

In that case, should we also update the `no-multiple-api-calls` to only look for `token.value === 'API'`? Because now it would flag the old api too.
https://github.com/Expensify/eslint-config-expensify/blob/9a56a5124e2a5340a020283a513d591381dcf852/eslint-plugin-expensify/no-multiple-api-calls.js#L11

**Comment by mountiny** [comment]
Renamed the files to the new agreed name and also I have omitted the `DeprecatedAPI` from the rule so this would only apply for actions and API commands.

Thanks for another look!

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js]: Yeah, that would make sense to me :+1:

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 4 above)
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js]: To confirm we're talking about this example yes?
```JS
        {
            // Make sure that we will not test a file which is not in '/actions/'
            code: `
                import Action from './src/libs/actions/Action.js'
                function test() {
                    API.write('Report_AddComment', params);
                    Action(params);
                }
            `,
            filename: './src/libs/notActions/Test.js',
        },
```

Yes, that should be invalid.

Comment (mountiny) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js]: Good yeah, that one is valid, because it is not in the `/actions/` directory

But we have an invalid which is as you described so I guess that is good.
```
{
            // Making an API call and calling an Action should be invalid
            code: `
                import Action from './src/libs/actions/Action.js'
                function test() {
                    API.write('Report_AddComment', params);
                    Action(params);
                }
            `,
            errors: [{
                message,
            }],
            filename: './src/libs/actions/Test.js',
        },
```

**Comment by mountiny** [comment]
Updated after Yuwen's comments, I think this is ready for another view @marcaaron and @yuwenmemon. Thanks!

File: eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js (hunk 4 above)
Comment (yuwenmemon) [reply eslint-plugin-expensify/tests/no-api-side-effect-actions.test.js]: Ah okay, thanks for clarifying!

**Review by yuwenmemon — APPROVED** [review]

**Comment by yuwenmemon** [comment]
@marcaaron all yours!

File: eslint-plugin-expensify/CONST.js [hunk 9]
```diff
@@ -13,8 +13,8 @@ module.exports = {
         PREFER_STR_METHOD: 'Prefer \'Str.{{method}}\' over the native function.',
 
         // eslint-disable-next-line max-len
-        NO_MULTIPLE_API_CALLS: 'Do not call API or DeprecatedAPI multiple times in the same method. The API response should return all the necessary data in a single request, and API calls should not be chained together.',
-
+        NO_MULTIPLE_API_CALLS: 'Do not call API multiple times in the same method. The API response should return all the necessary data in a single request, and API calls should not be chained together.',
+        NO_CALL_ACTIONS_FROM_ACTIONS: 'Calling actions from inside other actions is forbidden. If an action needs to call another action combine the two actions into a singular API call instead.',
```
Comment (marcaaron) [inline eslint-plugin-expensify/CONST.js:17]: Still need to test this against `Expensify/App` - but is this going to be a pain? We have a lot of "actions" called from "other actions" because the actions are not well defined.

**Review by marcaaron — APPROVED** [review]
